### ISSUE TITLE
"aws_iam_policy_document" Data Resource failing via JSON syntax

### ISSUE THREAD

**Issue by baskaran-md** [hashicorp/terraform#13911]
The Data resource is not accepting the JSON format specified in the doc - https://www.terraform.io/docs/configuration/syntax.html

### Terraform Version
Terraform v0.9.3

### Affected Resource(s)
- aws_iam_policy_document

### Terraform Configuration Files
https://gist.github.com/baskaran-md/08729036dbbcd14122e66674b2ec54c0

### Debug Output
https://gist.github.com/baskaran-md/c14271aadc1bb757e9f4d31dc771dac8

### References
- https://github.com/hashicorp/terraform/issues/13037

Even after fiddling with the json as per the Issue - #13037, it complaints in the actual aws iam resource, saying the reference to the data variable is unknown.

**Comment by nfroidure** [comment]
Having the same issue here with terraform v0.9.6, the following file fails saying `Failed to load root config module: Error loading terraform/test2/main.tf.json: position -: 'data' must be followed by exactly two strings: a type and a name`:

```
{
  "data": {
    "aws_iam_policy_document": {
      "lambda_assumed_role_policy": {
        "statement": {
          "actions": ["sts:AssumeRole"],
          "effect": "Allow",
          "principals": {
            "type": "Service",
            "identifiers": [
              "lambda.amazonaws.com",
              "apigateway.amazonaws.com",
              "events.amazonaws.com",
              "logs.amazonaws.com"
            ]
          }
        }
      },
      "lambda_role_policy": {
        "statement": {
          "resources": [
            "arn:aws:logs:*:*:*"
          ],
          "actions": [
            "logs:CreateLogGroup",
            "logs:CreateLogStream",
            "logs:PutLogEvents",
            "logs:DescribeLogStreams"
          ],
          "effect": "Allow"
        }
      }
    }
  }
}

```

While the exact same file with the `.tf` extension works:
```

data "aws_iam_policy_document" "lambda_assumed_role_policy" {
  statement {
    actions = ["sts:AssumeRole"]
    effect = "Allow"
    principals {
      type        = "Service"
      identifiers = [
        "lambda.amazonaws.com",
        "apigateway.amazonaws.com",
        "events.amazonaws.com",
        "logs.amazonaws.com"
      ]
    }
  }
}

data "aws_iam_policy_document" "lambda_role_policy" {
  statement {
    resources = [
      "arn:aws:logs:*:*:*"
    ]
    actions = [
      "logs:CreateLogGroup",
      "logs:CreateLogStream",
      "logs:PutLogEvents",
      "logs:DescribeLogStreams"
    ]
    effect = "Allow"
  }
}

```

Since the files are generated dynamically I really need to use JSON as input. Any idea when I can expect it to be fixed?

**Comment by nfroidure** [comment]
Seems related: https://github.com/hashicorp/terraform/issues/15466

**Comment by nfroidure** [comment]
Upgrading to 0.9.11 did not fix the issue.

**Comment by apparentlymart** [comment]
Hi @baskaran-md and @nfroidure! Sorry for this issue, and for the delayed response.

Unfortunately today there are some ambiguities in the JSON parsing that cause some strange behaviors like this. We're planning to revamp the JSON parsing soon to make it more "rigid" and well-defined so that it is clearer how to translate a HCL-based configuration to a JSON-based configuration and so that the parser can generate more helpful error messages if things aren't parsed correctly.

In the mean time, I'm not totally sure what to suggest here but I have worked around this sort of weirdness in the past by introducing arrays into the structure to help the parser understand how to partition things. For example:

```json
{
  "data": {
    "aws_iam_policy_document": {
      "lambda_assumed_role_policy": [
         {
           "statement": {
             "actions": ["sts:AssumeRole"],
             "effect": "Allow",
             "principals": {
               "type": "Service",
               "identifiers": [
                 "lambda.amazonaws.com",
                 "apigateway.amazonaws.com",
                 "events.amazonaws.com",
                 "logs.amazonaws.com"
               ]
             }
           }
         }
      ],
      "lambda_role_policy": [
         {
           "statement": {
             "resources": [
               "arn:aws:logs:*:*:*"
             ],
             "actions": [
               "logs:CreateLogGroup",
               "logs:CreateLogStream",
               "logs:PutLogEvents",
               "logs:DescribeLogStreams"
             ],
             "effect": "Allow"
           }
         }
      ]
    }
  }
}
```

**Comment by nfroidure** [comment]
@apparentlymart thanks for answering. The above file does not work either, like @baskaran-md I tried to put arrays everywhere possible with no luck. I finally found a workaround by simply templating raw `.tf` files instead of generating a JSON in the meanwhile.

Keep up the good work, Terraform is awesome ;).

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

I just tried with v0.12.0-alpha1 the JSON example from my most recent comment, and got the expected result this time:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": [
          "logs.amazonaws.com",
          "lambda.amazonaws.com",
          "events.amazonaws.com",
          "apigateway.amazonaws.com"
        ]
      }
    }
  ]
}
```

I also tried the version from @nfroidure's first comment above and got the same result.

With that said, it looks like the issue is fixed in alpha1, and so will be fixed in the forthcoming v0.12.0 final release, and so I'm going to close this out. Thanks for reporting this, and thanks for your patience while we laid the groundwork to get this working as expected.

Once v0.12.0 final is released, the website will include a new page with [full documentation on the JSON variant of the language](https://github.com/hashicorp/terraform/blob/009534d1016cd3b1d0d911f151520b04f8db453a/website/docs/configuration/syntax-json.html.md), showing how each construct from the native syntax maps into JSON. This, along with the redesigned JSON-based language parser, should mean an end to these quirky behaviors when input files are JSON.
